File: year2/semester2/mpp/java/TripClientServer/TripServer/src/main/java/net/TripClientRpcWorker.java
package net;

import domain.Account;
import domain.Booking;
import domain.Trip;
import dto.AccountDto;
import dto.BookingDto;
import dto.TripDto;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import service.TripObserver;
import service.TripServices;

import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.net.Socket;
import java.util.List;
import java.util.stream.Collectors;


public class TripClientRpcWorker implements Runnable, TripObserver {
    private TripServices service;
    private Socket connection;

    private ObjectInputStream input;
    private ObjectOutputStream output;
    private volatile boolean connected;
    private static final Logger LOGGER = LogManager.getLogger(TripClientRpcWorker.class.getName());

    public TripClientRpcWorker(TripServices tripServices, Socket connection) {
        LOGGER.traceEntry("initializing worker");
        this.service = tripServices;
        this.connection = connection;
        try {
            output = new ObjectOutputStream(connection.getOutputStream());
            output.flush();
            input = new ObjectInputStream(connection.getInputStream());
            connected = true;
        } catch (IOException e) {
            LOGGER.warn("initializing worker failed");
            e.printStackTrace();
        }
    }

    public void run() {
        LOGGER.traceEntry("running worker");
        while (connected) {
            try {
                Object request = input.readObject();
                LOGGER.info("received request " + request);
                Response response = handleRequest((Request) request);
                LOGGER.info("response for request " + response);
                if (response != null) {
                    LOGGER.info("sending response " + response);
                    sendResponse(response);
                }
            } catch (Exception e) {
                if (connected) {
                    LOGGER.warn("reading request failed");
                    e.printStackTrace();
                }
                break;
            }
        }
        try {
            LOGGER.info("closing connection");
            input.close();
            output.close();
            connection.close();
        } catch (IOException e) {
            LOGGER.warn("closing connection failed");
            e.printStackTrace();
        }
    }


    private static Response okResponse = new Response.Builder().type(ResponseType.OK).build();

    private static Response errorResponse = new Response.Builder().type(ResponseType.ERROR).build();

    private Response handleRequest(Request request) {
        LOGGER.info("handling request " + request);
        if (request.type() == RequestType.LOGIN) {
            LOGGER.info("handling LOGIN request");
            AccountDto accountDto = (AccountDto) request.data();
            try {
                LOGGER.info("finding account with name " + accountDto.name + "password " + accountDto.password);
                Account a = service.getAccountByNamePassword(accountDto.name, accountDto.password);
                LOGGER.info("found account " + a);
                if (a == null) {
                    return okResponse;
                }
                service.addTripObserver(this);
                return new Response.Builder()
                        .type(ResponseType.OK)
                        .data(new AccountDto(a.getId(), a.getName(), a.getPassword()))
                        .build();
            } catch (Exception e) {
                LOGGER.warn("finding account failed");
                return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
            }
        }
        if (request.type() == RequestType.GET_ALL_TRIPS) {
            LOGGER.info("handling GET_ALL_TRIPS request");
            try {
                List<Trip> trips = service.getAllTrips();
                List<TripDto> tripsDto = trips.stream()
                        .map(t -> new TripDto(t.getId(), t.getLandmark(), t.getCompanyName(), t.getDepartureTime(), t.getPrice(), t.getAvailablePlaces()))
                        .collect(Collectors.toList());
                LOGGER.info("returning response with trips");
                return new Response.Builder()
                        .type(ResponseType.OK)
                        .data(tripsDto)
                        .build();
            } catch (Exception e) {
                LOGGER.warn("getting all trips failed");
                return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
            }
        }
        if (request.type() == RequestType.LOGOUT) {
            LOGGER.info("Logout request");
            //service.logOut();
            service.removeTripObserver(this);
            connected = false;
            return okResponse;
        }
        if (request.type() == RequestType.SEARCH_TRIPS) {
            try {
                LOGGER.info("handling SEARCH_TRIPS request");
                TripDto tripDto = (TripDto) request.data();
                List<Trip> trips = service.getTripsByLandmarkDepartureHour(tripDto.landmark, tripDto.interval.get(0), tripDto.interval.get(1));
                List<TripDto> tripsDto = trips.stream()
                        .map(t -> new TripDto(t.getId(), t.getLandmark(), t.getCompanyName(), t.getDepartureTime(), t.getPrice(), t.getAvailablePlaces()))
                        .collect(Collectors.toList());
                LOGGER.info("returning response with trips searched");
                return new Response.Builder()
                        .type(ResponseType.OK)
                        .data(tripsDto)
                        .build();
            } catch (Exception e) {
                LOGGER.warn("getting searched trips failed");
                return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
            }
        }
        if (request.type() == RequestType.ADD_BOOKING) {
            LOGGER.info("handling ADD_BOOKING request");
            try {
                BookingDto bookingDto = (BookingDto) request.data();
                TripDto t = bookingDto.trip;
                AccountDto a = bookingDto.account;
                service.addBooking(bookingDto.client, bookingDto.phone, bookingDto.numTickets,
                        new Trip(t.id, t.landmark, t.companyName, t.departureTime, t.price, t.availablePlaces),
                        new Account(a.id, a.name, a.password));
                LOGGER.info("saved booking returning ok response");
                return okResponse;
            } catch (Exception ve) {
                LOGGER.warn("saving booking failed");
                return new Response.Builder().type(ResponseType.ERROR).data(ve.getMessage()).build();
            }
        }
        LOGGER.traceExit();
        return null;
    }

    private void sendResponse(Response response) throws IOException {
        LOGGER.info("sending response to client " + response);
        output.writeObject(response);
        output.flush();
    }

    @Override
    public void bookingInserted(Booking booking) {
        LOGGER.info("sending booking inserted response");
        BookingDto bookingDto = new BookingDto(booking.getId(), booking.getClientName(), booking.getPhoneNumber(),
                booking.getNumTickets(), new TripDto(booking.getTripId()), new AccountDto(booking.getAccountId()));
        Response response = new Response.Builder().type(ResponseType.BOOKING_ADDED).data(bookingDto).build();
        try {
            sendResponse(response);
        } catch (IOException e) {
            LOGGER.info("sending response booking inserted failed");
            e.printStackTrace();
        }
    }
}

File: year2/semester2/mpp/c#/TripClientServer/TripsServer/Net/TripClientRpcWorker.cs
using log4net;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Sockets;
using System.Runtime.Serialization;
using System.Runtime.Serialization.Formatters.Binary;
using TripCommon.domain;
using TripCommon.Dto;
using TripCommon.Net;
using TripCommon.Validator;

namespace TripsServer.Net
{
    class TripClientRpcWorker: ITripObserver
    {
        private static readonly ILog LOGGER = LogManager.GetLogger("TripClientRpcWorker");

        private ITripServices tripServices;
        private TcpClient connection;

        private NetworkStream stream;
        private IFormatter formatter;
        private volatile bool connected;
        public TripClientRpcWorker(ITripServices tripServices, TcpClient connection)
        {
            LOGGER.Info("initializing worker");
            this.tripServices = tripServices;
            this.connection = connection;
            try
            {
                stream = connection.GetStream();
                formatter = new BinaryFormatter();
                connected = true;
            }
            catch (Exception e)
            {
                LOGGER.Warn("initializing worker failed" + e);
                Console.WriteLine(e.StackTrace);
            }
        }

        public void BookingInserted(Booking booking)
        {
            LOGGER.InfoFormat("booking inserted {0}", booking);
            BookingDto bookingDto = new BookingDto(booking.Id, booking.ClientName, booking.PhoneNumber, booking.NumTickets,
                new TripDto(booking.TripId), new AccountDto(booking.AccountId));
            Response response = new Response()
            {
                Type = ResponseType.BOOKING_ADDED,
                Data = bookingDto
            };
            try
            {
                SendResponse(response);
            }
            catch(Exception e)
            {
                LOGGER.Warn("sending response booking inserted failed");
                LOGGER.Warn(e.StackTrace);
            }
        }

        public void Run()
        {
            LOGGER.Info("running worker");
            while (connected)
            {
                try
                {
                    Request request = (Request)formatter.Deserialize(stream);
                    LOGGER.InfoFormat("received request {0}", request);
                    Response response = HandleRequest(request);
                    LOGGER.InfoFormat("response for request {0}", response);
                    if (response != null)
                    {
                        LOGGER.InfoFormat("sending response to client {0}", response);
                        SendResponse(response);
                    }
                }
                catch (Exception e)
                {
                    if (connected)
                    {
                        LOGGER.Warn("running worker stopped with exception {0}", e);
                        LOGGER.Warn(e.StackTrace);
                    }
                    break;
                }
            }
            LOGGER.Info("closing connection");
            try
            {
                stream.Close();
                connection.Close();
            }
            catch (Exception e)
            {
                LOGGER.Warn("Error closing connection" + e);
                LOGGER.Warn(e.StackTrace);
            }
        }
        private Response HandleRequest(Request request)
        {
            LOGGER.Info("handling request " + request);
            if (request.Type == RequestType.LOGIN)
            {
                LOGGER.Info("Login request");
                AccountDto accountDto = (AccountDto)request.Data;
                Account account = null;
                try
                {
                    lock (tripServices)
                    {
                        account = tripServices.FindAccountByNameAndPassword(accountDto.name, accountDto.password);
                    }
                    if (account == null)
                    {
                        return new Response()
                        {
                            Type = ResponseType.OK
                        };
                    }
                    tripServices.AddTripObserver(this);
                    return new Response()
                    {
                        Type = ResponseType.OK,
                        Data = new AccountDto(account.Id, account.Name, account.Password)
                    };
                }
                catch (Exception e)
                {
                    LOGGER.Warn("handling login failed ");
                    return new Response()
                    {
                        Type = ResponseType.ERROR,
                        Data = e.Message
                    };
                }
            }
            if (request.Type == RequestType.LOGOUT)
            {
                LOGGER.Info("Logout request");
                lock (tripServices)
                {
                    tripServices.Logout();
                }
                tripServices.RemoveTripObserver(this);
                connected = false;
                return new Response()
                {
                    Type = ResponseType.OK
                };
            }
            if (request.Type == RequestType.GET_ALL_TRIPS)
            {
                LOGGER.Info("get all trips request");
                List<Trip> trips = null;
                try
                {
                    lock (tripServices)
                    {
                        trips = tripServices.GetAllTrips();
                    }
                    return new Response()
                    {
                        Type = ResponseType.OK,
                        Data = trips
                                    .Select(t => new TripDto(t.Id, t.Landmark, t.CompanyName, t.DepartureTime, t.Price, t.AvailablePlaces))
                                    .ToList()
                    };
                }
                catch (Exception e)
                {
                    LOGGER.Warn("getting all trips failed ");
                    return new Response()
                    {
                        Type = ResponseType.ERROR,
                        Data = e.Message
                    };
                }
            }
            if (request.Type == RequestType.SEARCH_TRIPS)
            {
                LOGGER.Info("search trips request");
                List<Trip> trips = null;
                try
                {
                    TripDto tripDto = (TripDto)request.Data;
                    lock (tripServices)
                    {
                        trips = tripServices.GetTripsByLandmarkDepartureHour(tripDto.landmark, tripDto.start, tripDto.end);
                    }
                    LOGGER.Info("returning response with searched trips");
                    return new Response()
                    {
                        Type = ResponseType.OK,
                        Data = trips
                                    .Select(t => new TripDto(t.Id, t.Landmark, t.CompanyName, t.DepartureTime, t.Price, t.AvailablePlaces))
                                    .ToList()
                    };
                }
                catch (Exception e)
                {
                    LOGGER.Warn("searching trips failed ");
                    return new Response()
                    {
                        Type = ResponseType.ERROR,
                        Data = e.Message
                    };
                }
            }
            if (request.Type == RequestType.ADD_BOOKING)
            {
                LOGGER.Info("add booking request");
                BookingDto bookingDto = (BookingDto)request.Data;
                TripDto tripDto = bookingDto.trip;
                AccountDto accountDto = bookingDto.account;
                Trip t = new Trip(tripDto.id, tripDto.landmark, tripDto.companyName, tripDto.departure, tripDto.price, tripDto.places);
                Account a = new Account(accountDto.id, accountDto.name, accountDto.password);
                try
                {
                    lock (tripServices)
                    {
                        tripServices.AddBooking(bookingDto.client, bookingDto.phone, bookingDto.numTickets, t, a);
                    }
                    LOGGER.Info("saved booking returning ok response");
                    return new Response()
                    {
                        Type = ResponseType.OK
                    };
                }
                catch (Exception e)
                {
                    LOGGER.Warn("add booking failed ");
                    return new Response()
                    {
                        Type = ResponseType.ERROR,
                        Data = e.Message
                    };
                }
            }
            return null;
        }
        private void SendResponse(Response response)
        {
            LOGGER.InfoFormat("sending response {0}", response);
            formatter.Serialize(stream, response);
            stream.Flush();
        }
    }
}
